### PR TITLE
PP-602: pidcache to cull list of processes to poll in MOM polling is broken by PID recycling; pidcache_truncate doesn't allow for new kthreads

### PR DESCRIPTION
**Pull request by ashwathraop**
#### Issue-ID

* **[PP-602](https://pbspro.atlassian.net/browse/PP-602)**

#### Problem description
* PID recycling can break the pidcache created to avoid MOM polling of non-job processes.

#### Cause / Analysis
* We maintain a pidcache which includes all the system processes and kernel threads that we want pbs mom to ignore during polling. This cache is created when mom starts and referred on every poll and any process that does not exist anymore is deleted from the cache. When a process added to this pidcache ends and same pid is assigned to a job process (say after pid recycle) before a new poll, during next poll mom sees this pid  still alive and does not bother to delete the same from the cache. As of now there is no validation of processes to see if they actually changed since they are added to cache.

#### Solution description
* This pidcache code is enabled only on sgi_uv cpuset machines. So the fix will cache only those processes we want to PBS to monitor. That’s basically list of processes fetched from /dev/cpuset/<jobid>/tasks files. Once we have this cache we will go through /proc directory and fetch details of only cached processes. So with this approach we are creating and destroying cache for every mom_get_sample. Creating a cache for every mom_get_sample isn’t very much expensive as we are looking only at /dev/cpuset tasks files. This will also makes sure we are not skipping any new job process which got its pid from a previous kthread or system process which died amid of two mom_get_sample calls.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/resmom/linux/mom_mach.c [hunk 1]
```diff
@@ -129,13 +129,16 @@ static char	procfs[] = "/proc";
 static DIR	*pdir = NULL;
 static int	pagesize;
 static long	hz;
+#if	MOM_CPUSET
+static char cpusetfs[] = "/dev/cpuset/PBSPro";
```
Comment (mike0042) [inline src/resmom/linux/mom_mach.c]: You should include mom_mach.h and use the PBS_CPUSETDIR macro here.

Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Fixed it. We don't need to import mom_mach.h again as it is included under mom_func.h

File: src/resmom/linux/mom_mach.c [hunk 2]
```diff
@@ -8001,6 +8008,66 @@ pidcache_create(void)
...
+ * @par MT-Safe:	yes
+ * @par Side Effects:
+ *	None
+ */
+static int
+pidcache_reset(pidcachetype_t *pidcache)
+{
+	unsigned long long	numerator, denominator;
+	unsigned long		setsize;
+	pidcachetype_t		*new_set;
```
Comment (mike0042) [inline src/resmom/linux/mom_mach.c]: Looks like some unused variables are being declared that will likely result in compiler warnings. Please ensure your build log is free of warnings. There is only one expected warning related to ical.h.

Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: @mike0042 Yes these variables are not needed. Removed them. Please check now.

File: src/resmom/linux/mom_mach.c [hunk 3]
```diff
@@ -8001,6 +8008,63 @@ pidcache_create(void)
...
+			if (pidcache_insert(pid, pidcache))
+				nprocs++;
+		}
+		fclose(fd);
+	}
+
+	if (ncantstat)
+		return -1;
+	else
+		return nprocs;
```
Comment (prakashcv13) [inline src/resmom/linux/mom_mach.c]: I believe we should always return nprocs if (nprocs > 0). The return value is used to log the number of pids that are cached. 
We should return -1 only if we do not add any pid to the cache...
With the current logic even if we have added pids to the cache, we would return -1, and report ncached to be zero...

File: src/resmom/linux/mom_mach.c [hunk 4]
```diff
@@ -8001,6 +8008,63 @@ pidcache_create(void)
 	return (pidcache_arena);
```
Comment (prakashcv13) [inline src/resmom/linux/mom_mach.c]: not related to the current fix, but we do use pidcache_create(), shall we check for the return value of calloc() and set/not set pidcache_starttime accordingly...

File: src/resmom/linux/mom_mach.c [hunk 5]
```diff
@@ -222,20 +225,21 @@ pbs_plinks	*Proc_lnks = NULL;	/* process links table head */
...
+#if	MOM_CPUSET
 typedef int		pidcachetype_t;		/* type of allocatable unit */
 static pidcachetype_t	*pidcache_arena;	/* cache storage area */
 static unsigned int 	pidcache_bitsper = sizeof(pidcachetype_t) * NBBY;
 static int		pidcache_check(pid_t, pidcachetype_t *);
 static pidcachetype_t	*pidcache_create(void);
+static int		pidcache_reset(pidcachetype_t *);
 static void		pidcache_destroy(void);
-static int		pidcache_eligible(proc_stat_t *, char *, char *, unsigned long long);
 static pidcachetype_t *	pidcache_getarena(void);
```
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Indentation looks fine to me. I am not sure which line you are referring to.

File: src/resmom/linux/mom_mach.c [hunk 6]
```diff
@@ -5534,8 +5538,9 @@ mom_get_sample(void)
 	char			procname[256];
 	struct stat		sb;
 	proc_stat_t		*ps;
-	int			maxexcludedPID = 0;
+#if MOM_CPUSET
 	pidcachetype_t		*pidcache;
```
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Will fix it.

File: src/resmom/linux/mom_mach.c (hunk 4 above)
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Looking at the comment I just realized we don't need pidcache_starttime anymore. On calloc failure we can return from mom_get_sample() itself.

File: src/resmom/linux/mom_mach.c [hunk 7]
```diff
@@ -8001,6 +8008,63 @@ pidcache_create(void)
...
+	struct dirent		*dent;
+	pid_t			pid;
+	char			taskname[256];
+	int			ncantstat = 0;
+	FILE			*fd;
+	char			line[25] = {0};
+	char			*p;
+	int			i;
+	int			nprocs = 0;
+
```
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Will do.

File: src/resmom/linux/mom_mach.c (hunk 3 above)
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Yes.

**Comment by ashwathraop** [comment]
@prakashcv13 Please have a look at the code now.

File: src/resmom/linux/mom_mach.c (hunk 5 above)
Comment (prakashcv13) [inline src/resmom/linux/mom_mach.c]: screenshot from 2017-02-14 15-36-57
this is how I am seeing it..

File: src/resmom/linux/mom_mach.c [hunk 8]
```diff
@@ -5548,16 +5547,18 @@ mom_get_sample(void)
...
 	if (pdir == NULL)
 		return PBSE_INTERNAL;
 
 #if MOM_CPUSET
 	if (((pidcache = pidcache_getarena()) == NULL) && pidcache_needed()) {
-		if ((pidcache = pidcache_create()) == NULL)
+		if ((pidcache = pidcache_create()) == NULL) {
 			log_err(errno, __func__, "PID cache create");
+			return PBSE_SYSTEM;
+		}
```
Comment (prakashcv13) [inline src/resmom/linux/mom_mach.c]: previously, we were not returning if pidcache_create() failed. Now, we are. Why so?

File: src/resmom/linux/mom_mach.c [hunk 9]
```diff
@@ -8059,7 +8059,7 @@ pidcache_reset(pidcachetype_t *pidcache)
 		fclose(fd);
 	}
 
-	if (ncantstat)
+	if (!nprocs && ncantstat)
```
Comment (prakashcv13) [inline src/resmom/linux/mom_mach.c]: I would like to understand what would happen if there were no entries in the cpusetdir. In that case, should we be returning -1?

File: src/resmom/linux/mom_mach.c [hunk 10]
```diff
@@ -8022,14 +8022,14 @@ pidcache_create(void)
 static int
 pidcache_reset(pidcachetype_t *pidcache)
 {
-	struct dirent		*dent;
-	pid_t			pid;
+	struct dirent		*dent = NULL;
+	pid_t			pid = 0;
 	char			taskname[256];
```
Comment (prakashcv13) [inline src/resmom/linux/mom_mach.c]: should we be initializing task_name as well?

Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: Okay.

File: src/resmom/linux/mom_mach.c (hunk 9 above)
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: No. It should return 0. Its not an error case not finding any job processes. mom will not have any processes to monitor and function should return successfully.

File: src/resmom/linux/mom_mach.c (hunk 5 above)
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: I am not sure why it is showing like thin in diff mode. Please have a look with "view the whole file" and you should see the indentation right.

File: src/resmom/linux/mom_mach.c (hunk 8 above)
Comment (ashwathraop) [inline src/resmom/linux/mom_mach.c]: I added this check to validate calloc failure that could happen within pidcache_create(). Now revisiting the code I see we already safeguarded all our pidcache usage code with a (pidcache != NULL) check. So by removing this return even on calloc failure we will continue with considering all the processes running the system to be monitored by MoM as we used to do earlier.

**Comment by ashwathraop** [comment]
@prakashcv13 I have made the changes. Have a look.

**Pull request merged**